crates/lib/src/execution_data.rs

```diff
@@ -1,120 +1,334 @@
-use std::collections::HashMap;
-use std::convert::TryFrom;
-use std::num::{ParseIntError, TryFromIntError};
+use enum_as_inner::EnumAsInner;
+use num::complex::Complex64;
+use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, HashMap};
+use std::num::ParseIntError;
 use std::str::FromStr;
 use std::time::Duration;
 
-use qcs_api_client_grpc::models::controller::ReadoutValues;
-use quil_rs::instruction::MemoryReference;
+use itertools::Itertools;
+use ndarray::prelude::*;
 
-use crate::RegisterData;
+use crate::{
+    qpu::result_data::{QpuResultData, ReadoutValues},
+    qvm::QvmResultData,
+    RegisterData,
+};
 
-/// The result of executing an [`Executable`](crate::Executable) via
-/// [`Executable::execute_on_qvm`](crate::Executable::execute_on_qvm).
+/// Represents the two possible types of data returned from either the QVM or a real QPU.
+/// Each variant contains the original data returned from its respective executor.
+///
+/// # Usage
+///
+/// Your usage of [`ResultData`] will depend on the types of programs you are running and where.
+/// The `to_register_map()` method will attempt to build a [`RegisterMap`] out of the data, where each
+/// register name is mapped to a 2-dimensional rectangular [`RegisterMatrix`] where each row
+/// represents the final values in each register index for a particular shot. This is often the
+/// desired form of the data and it is _probably_ what you want. This transformation isn't always
+/// possible, in which case `to_register_map()` will return an error.
+///
+/// To understand why this transformation can fail, we need to understand a bit about how readout data is
+/// returned from the QVM and from a real QPU:
+///
+/// The QVM treats each `DECLARE` statement as initialzing some amount of memory. This memory works
+/// as one might expect it to. It is zero-initalized, and subsequent writes to the same region
```

Comment:
```suggestion
/// as one might expect it to. It is zero-initialized on each shot, and subsequent writes to the same region
```

Reply:
Fair, I'll commit this, though there is an [open QVM issue](https://github.com/quil-lang/qvm/issues/313) related to this.